include/circt/Support/SymUniquer.h

```diff
@@ -0,0 +1,76 @@
+//===- SymUniquer.h - Declare Symbol Uniquer --------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file declares a Symbol Uniquer.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_SUPPORT_SYMUNIQUER_H
+#define CIRCT_SUPPORT_SYMUNIQUER_H
+
+#include "circt/Support/SymCache.h"
+
+namespace circt {
+
+/// The SymbolUniquer is used to generate unique names based on some set of
+/// existing names defined through a SymbolCache.
+class SymbolUniquer {
+  /// Uniquer function which suffixes a base name with an integer, starting at
+  /// 0.
+  static std::pair<std::string, llvm::Any>
+  integerUniquer(mlir::StringRef name, llvm::Any previousToken) {
```

Comment:
I'll give this a more thorough review once the symbol cache stuff is moved and is divorced from this PR. Early comment: why `llvm::Any` vs a template? I am wary of any sort of dynamic type behavior.

Reply:
This design was motivated by not having `SymbolUniquer` be a template, but I do see your point of view. A templated `SymbolUniquer` with a default implementation being "integer-uniqueing" (as shown) could be an alternative.